### DISCUSSION TITLE
api-analysis plugin?

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-tycho/tycho#1233] [Ideas]
It seems currently APi analysis is implemented as a combination of `maven-antrun-plugin`, `tycho-dependency-tools-plugin` and `eclipse-run-plugin` firing up a separate process, this feels hard to setup, understand maintain and possibly influence performance (e.g. API target needs to evaluated multiple times), I wonder why Tycho do not include a `tycho-api-analysis` plugin that performs all necessary steps and even call the application inside the build jvm to prevent the additional cold process creation penalty?

**Comment by akurtakov** [comment]
Answer is easy -  no one did so :) . Back in the day it was run only for I-builds totally decoupled from verification builds, it was kdaniel and me that added first "hacky" implementation to cbi-plugins to generate build.xml file per project and call it after that with maven-antrun-plugin, @mickaelistria added the application in pde and changed to the far simpler build we have now (which is not simple and straightforward at all) and etc.  Overall, progress with baby steps.

**Comment by HannesWell** [comment]
I would appreciate such plugin very much and already thought about it multiple times. Isn't ist similar to https://github.com/eclipse/tycho/issues/115?

As you said it would be more convenient and also offers posible performance improvements. The baseline-target resolution for each plugin could be improved as well. E.g. it could be resolved only once for all Plugins in the reactor instead of multiple times for each plugin. Maybe we could also leverage Tychos TP resolution?

But besides the analysis such plugin should also generate the API-description, probably as separate goal. So that we have `tycho-api-plugin:analyze` and `tycho-api-plugin:generate-description` or similar. Ideally those Goals are bound automatically to a suitable phase so that a user only has to add the Plugin to the build and only needs to configure the Baseline p2 repo.

**Reply by laeubi** [reply]
#115 is about a real baseline plugin based on BND / OSGI semantic version. I'll try to create a basice api-tools plugin for tycho we can build on but would need probably some help with creating some useful integration tests later on.

**Reply by HannesWell** [reply]
What exactly would be the difference besides the technology under the hood? As far as I understand it the purpose would be the same?

**Comment by laeubi** [comment]
I have looked into it and thing at least we can cache some content, **but** currently the API analysis is implemented as a builder that requires a lot of work to be done for setup, so one can not really pass the compiled code and/or sources here, that's a bit limiting at the moment.
Beside this, one might choose a similar approach as we did with the `tycho-bundles-product` that PDE provides an ApiToolsProduct that Tycho can reference here and Tycho installs a communicator bundle into that product, then we would not need to create a new installation on each build and we can start one instance that might server as some kind of Api-Tools server ...

**Reply by HannesWell** [reply]
Sounds interesting.

As very first step I suggest to replicate the steps currently done in the Platform-aggregator build api-check profile in an api-check Mojo.
Afterwards we can then improve the implementation step by step. But we would already gain conveniens which would likely increase the Number of adopters.

Furthermore we could then consider to drop/deprecate-for-removal the `list-dependencies` Mojo in Tycho as well as the `APIAnalysisApplication` in PDE since they were only created for the aggregator profile. I think we can maintain that better If everything is done internally in Tycho and not split over multiple Mojos and Projects.

I can provide the first implementation outlined above when back from vacation, unless you want to do it differently.

**Reply by HannesWell** [reply]
@laeubi are you currently working on the analysis mojo?
If not I would start to work on this at the Weekend with the suggested approach, so that we at least have a working Mojo for the Tycho 3.0 release.
We can then optimize it later on.

**Reply by laeubi** [reply]
Fell free to do so, but I don't think Tycho should provide the `APIAnalysisApplication` but PDE should provide a suitable product/package we can consume at Tycho...
